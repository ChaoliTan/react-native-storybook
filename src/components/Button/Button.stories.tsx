import React from 'react';
import { View } from 'react-native';
import { storiesOf } from '@storybook/react-native';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';
import Button from '.';

export default {
    title: '01 - Button',
    decorators: [withKnobs],
    component: Button,
};

const toggleGroupId = 'Toggles';
const valueGroupId = 'Values';

const defaultButtonProps = () => ({
    children: text('title', 'default button', valueGroupId),
    disabled: boolean('disabled', false, toggleGroupId),
    title: 'default button',
    onPress: () => {},
});

const disabledButtonProps = () => ({
    children: text('title', 'disabled button', valueGroupId),
    disabled: boolean('disabled', true, toggleGroupId),
    title: 'disabled button',
    onPress: () => {},
});

export const defaultButton = () => {
    const { children } = defaultButtonProps();
    return (
        <View style={{ width: 200, margin: 50 }}>
            <Button {...defaultButtonProps()}>{children}</Button>
        </View>
    );
};

export const disabledButton = () => {
    const { children } = disabledButtonProps();
    return (
        <View style={{ width: 200, margin: 50 }}>
            <Button {...disabledButtonProps()}>{children}</Button>
        </View>
    );
};
