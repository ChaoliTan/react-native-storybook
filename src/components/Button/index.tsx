import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import { useTheme } from '@/hooks';
import buttonStyle, { ButtonSize, ButtonStyles, ButtonType } from './style';

type ButtonProps = {
    size?: ButtonSize;
    type?: ButtonType;
    style?: ButtonStyles;
    loading?: boolean;
    disabled?: boolean;
    children: React.ReactText | React.ReactNode;
    onPress?: (e?: any) => void;
};

const ThemedButton: React.FC<ButtonProps> = (props) => {
    const { children, size = 'normal', type = 'default', style, loading, disabled, onPress, ...restProps } = props;
    const [isPressIn, setIsPresseIn] = React.useState(false);
    const _styles = useTheme<ButtonStyles>({ themeStyle: buttonStyle, style });

    const wrapperStyle = React.useMemo(
        () => [
            _styles.wrapper,
            _styles[`${size}_raw`],
            _styles[`${type}_raw`],
            disabled && _styles[`${type}_disabled_raw`],
            style,
        ],
        [_styles],
    );
    const textStyle = React.useMemo(
        () => [_styles[`${size}_rawText`], _styles[`${type}_rawText`], disabled && _styles[`${type}_disabled_rawText`]],
        [_styles, disabled],
    );

    return (
        <TouchableOpacity
            activeOpacity={0.85}
            style={wrapperStyle}
            disabled={disabled}
            onPressIn={() => setIsPresseIn(true)}
            onPress={(e) => onPress?.(e)}
            onPressOut={() => setIsPresseIn(false)}
            {...restProps}
        >
            <View style={_styles.container}>
                {/* TODO: add loading Indicator  */}
                {loading ? <Text style={textStyle}>loading...</Text> : <Text style={textStyle}>{children}</Text>}
            </View>
        </TouchableOpacity>
    );
};

export default ThemedButton;
