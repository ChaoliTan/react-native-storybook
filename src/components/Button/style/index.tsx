import { StyleSheet, TextStyle, ViewStyle } from 'react-native';
import { Theme } from '@/style';

export type ButtonSize = 'large' | 'normal' | 'small';
export type ButtonType = 'default' | 'primary' | 'plain';
/**
 * container: 容器
 * wrapper: children的容器
 * raw: 负责跟按钮 size/type 相关的盒模型样式
 * rawText: 负责跟按钮 size/type 相关的文字样式
 */
export interface ButtonStyles {
    container: ViewStyle;
    wrapper: ViewStyle;
    // size
    normal_raw: TextStyle;
    normal_rawText: TextStyle;

    // default style
    default_raw: ViewStyle;
    default_disabled_raw: ViewStyle;
    default_rawText: TextStyle;
    default_disabled_rawText: TextStyle;
}
export default (theme: Theme) =>
    StyleSheet.create<ButtonStyles>({
        container: {
            flexDirection: 'row',
        },
        wrapper: {
            alignItems: 'center',
            justifyContent: 'center',
            borderRadius: theme.radius_md,
            borderColor: theme.primary,
            borderWidth: 1,
        },
        // size
        normal_raw: {
            paddingVertical: theme.h_spacing_md,
        },
        normal_rawText: {},

        // default style
        default_raw: {},
        default_disabled_raw: {},
        default_rawText: {
            color: theme.secondary,
        },
        default_disabled_rawText: {},
    });
