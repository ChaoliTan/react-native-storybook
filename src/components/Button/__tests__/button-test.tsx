import 'react-native';
import React from 'react';
import Button from '../';
// 1. mock native modules
// 2. API mock
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';

// snapshot
// image snapshot
it('renders <Button/> correctly', () => {
    const button = renderer.create(<Button onPress={jest.fn()}>This is a button</Button>).toJSON();
    expect(button).toMatchSnapshot();
});

describe('<Button/> functioning correctly', () => {
    it('display text correctly', () => {
        const title = 'text title 123';
        const button = render(<Button onPress={jest.fn()}>{title}</Button>);
        expect(button.getByText(title)).toBeTruthy();
    });
    it('invoke onPress correctly', () => {
        const pressFn = jest.fn();
        const button = render(<Button onPress={pressFn}>test</Button>).container;
        fireEvent.press(button);
        expect(pressFn).toBeCalled();
    });
    it('should be disabled', () => {
        const pressFn = jest.fn();
        const button = render(
            <Button onPress={pressFn} disabled>
                test
            </Button>,
        ).container;
        fireEvent.press(button);
        // TODO mock TouchableOpacity onPress
        expect(pressFn).toBeCalledTimes(1);
    });
});
