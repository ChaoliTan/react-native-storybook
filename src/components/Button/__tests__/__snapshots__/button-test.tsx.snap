// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <Button/> correctly 1`] = `
<TouchableOpacity
  activeOpacity={0.8}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 5,
        "borderWidth": 1,
        "justifyContent": "center",
      },
      Object {
        "paddingVertical": 8,
      },
      Object {
        "borderColor": "#FF3C51",
      },
      undefined,
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        Array [
          Object {},
          Object {
            "color": "#FF5839",
          },
          undefined,
        ]
      }
    >
      This is a button
    </Text>
  </View>
</TouchableOpacity>
`;
